### PR TITLE
core: enhance error handling

### PR DESCRIPTION
**Pull request by satoru-takeuchi**
**Description of your changes:**

- handle errors in callers as possible
- enrich error messages

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] CI tests has been updated, if necessary.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.

### CODE REVIEW

File: pkg/dr/health.go [hunk 1]
```diff
@@ -64,16 +64,19 @@ func Health(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespa
...
 	}
 	logging.Info(cephStatus)
 
 	logging.Info("running mirroring daemon health")
 
-	exec.RunCommandInOperatorPod(ctx, clientsets, "rbd", []string{"-p", mirrorBlockPool.Name, "mirror", "pool", "status"}, cephClusterNamespace, operatorNamespace, false, false)
+	_, err = exec.RunCommandInOperatorPod(ctx, clientsets, "rbd", []string{"-p", mirrorBlockPool.Name, "mirror", "pool", "status"}, cephClusterNamespace, operatorNamespace, false)
+	if err != nil {
+		logging.Error(nil)
+	}
```
Comment (Madhu-1) [inline pkg/dr/health.go:79]: should be logging `err` here? if not this is of not much help

File: pkg/exec/exec.go [hunk 2]
```diff
@@ -37,70 +36,85 @@ var (
...
 
-	execCmdInPod(ctx, clientsets, cmd, pod.Name, "rook-ceph-operator", pod.Namespace, clusterNamespace, args, &stdout, &stderr, returnOutput, exitOnError)
+	err = execCmdInPod(ctx, clientsets, cmd, pod.Name, "rook-ceph-operator", pod.Namespace, clusterNamespace, args, &stdout, &stderr, returnOutput)
 	if !returnOutput {
-		return ""
+
+		return "", err
 	}
 
 	fmt.Print(stderr.String())
```
Comment (Madhu-1) [inline pkg/exec/exec.go:56]: dont we need to consider `stderr` for anything? currently its treated with no value except for printing

File: pkg/rook/purge_osd.go [hunk 3]
```diff
@@ -51,5 +53,9 @@ func PurgeOsd(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNames
 	}
 	logging.Info("Running purge osd command")
 
-	exec.RunCommandInOperatorPod(ctx, clientsets, cmd, args, operatorNamespace, clusterNamespace, false, true)
+	_, err = exec.RunCommandInOperatorPod(ctx, clientsets, cmd, args, operatorNamespace, clusterNamespace, false)
+	if err != nil {
+		logging.Error(err, "failed to remove osd %s", osdId)
+		os.Exit(1)
+	}
```
Comment (Madhu-1) [inline pkg/rook/purge_osd.go:60]: This block of code is repeated in many places may be a helper function for it helps?

File: cmd/commands/ceph.go [hunk 4]
```diff
@@ -33,6 +35,10 @@ var CephCmd = &cobra.Command{
 	},
 	Run: func(cmd *cobra.Command, args []string) {
 		logging.Info("running 'ceph' command with args: %v", args)
-		exec.RunCommandInOperatorPod(cmd.Context(), clientSets, cmd.Use, args, operatorNamespace, cephClusterNamespace, false, true)
+		_, err := exec.RunCommandInOperatorPod(cmd.Context(), clientSets, cmd.Use, args, operatorNamespace, cephClusterNamespace, false)
+		if err != nil {
+			logging.Error(err)
+			os.Exit(1)
```
Comment (subhamkrai) [inline cmd/commands/ceph.go:41]:
```suggestion
			logging.Fatal(err)
```
`logging.Fatal` is combination of what you have

File: cmd/commands/rbd.go [hunk 5]
```diff
@@ -31,6 +34,10 @@ var RbdCmd = &cobra.Command{
 		verifyOperatorPodIsRunning(cmd.Context(), clientSets)
 	},
 	Run: func(cmd *cobra.Command, args []string) {
-		exec.RunCommandInOperatorPod(cmd.Context(), clientSets, cmd.Use, args, operatorNamespace, cephClusterNamespace, false, true)
+		_, err := exec.RunCommandInOperatorPod(cmd.Context(), clientSets, cmd.Use, args, operatorNamespace, cephClusterNamespace, false)
+		if err != nil {
+			logging.Error(err)
+			os.Exit(1)
```
Comment (subhamkrai) [inline cmd/commands/rbd.go:40]:
```suggestion
			logging.Error(err)
			os.Exit(1)
```
same

File: cmd/commands/rook.go [hunk 6]
```diff
@@ -38,7 +39,11 @@ var versionCmd = &cobra.Command{
 	Short: "Prints rook version",
 	Args:  cobra.NoArgs,
 	Run: func(cmd *cobra.Command, _ []string) {
-		exec.RunCommandInOperatorPod(cmd.Context(), clientSets, "rook", []string{cmd.Use}, operatorNamespace, cephClusterNamespace, false, true)
+		_, err := exec.RunCommandInOperatorPod(cmd.Context(), clientSets, "rook", []string{cmd.Use}, operatorNamespace, cephClusterNamespace, false)
+		if err != nil {
+			logging.Error(err)
+			os.Exit(1)
```
Comment (subhamkrai) [inline cmd/commands/rook.go:45]:
```suggestion
			logging.Error(err)
			os.Exit(1)
```
same

File: cmd/commands/root.go [hunk 7]
```diff
@@ -123,7 +124,11 @@ func preValidationCheck(ctx context.Context, k8sclientset *k8sutil.Clientsets) {
 }
 
 func verifyOperatorPodIsRunning(ctx context.Context, k8sclientset *k8sutil.Clientsets) {
-	rookVersionOutput := exec.RunCommandInOperatorPod(ctx, k8sclientset, "rook", []string{"version"}, operatorNamespace, cephClusterNamespace, true, false)
+	rookVersionOutput, err := exec.RunCommandInOperatorPod(ctx, k8sclientset, "rook", []string{"version"}, operatorNamespace, cephClusterNamespace, true)
+	if err != nil {
+		logging.Error(err, "failed to get rook version")
+		os.Exit(1)
```
Comment (subhamkrai) [inline cmd/commands/root.go:130]:
```suggestion
		logging.Error(err, "failed to get rook version")
		os.Exit(1)
```

same

File: pkg/dr/health.go [hunk 8]
```diff
@@ -64,16 +64,19 @@ func Health(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespa
...
 		return
 	}
 	logging.Info(cephStatus)
 
 	logging.Info("running mirroring daemon health")
 
-	exec.RunCommandInOperatorPod(ctx, clientsets, "rbd", []string{"-p", mirrorBlockPool.Name, "mirror", "pool", "status"}, cephClusterNamespace, operatorNamespace, false, false)
+	_, err = exec.RunCommandInOperatorPod(ctx, clientsets, "rbd", []string{"-p", mirrorBlockPool.Name, "mirror", "pool", "status"}, cephClusterNamespace, operatorNamespace, false)
+	if err != nil {
+		logging.Error(nil)
```
Comment (subhamkrai) [inline pkg/dr/health.go:78]:
```suggestion
		logging.Fatal(err)
```
?

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: pkg/dr/health.go (hunk 1 above)
Comment (satoru-takeuchi) [reply pkg/dr/health.go:79]: Indeed, thanks.

File: pkg/exec/exec.go (hunk 2 above)
Comment (satoru-takeuchi) [reply pkg/exec/exec.go:56]: Wrapped `err` by `stderr` string. So caller can use it.

File: pkg/rook/purge_osd.go (hunk 3 above)
Comment (satoru-takeuchi) [reply pkg/rook/purge_osd.go]: Replaced with logging.Fatal() as @subhamkrai suggested.

File: cmd/commands/ceph.go (hunk 4 above)
Comment (satoru-takeuchi) [reply cmd/commands/ceph.go]: Yeah, I should have read the implementation of `logging.Fatal` carefully.

**Review by subhamkrai — CHANGES_REQUESTED** [review]
Please rebase your PR to get CI fixes. Thanks

**Comment by satoru-takeuchi** [comment]
OK, I'll fix this CI failure.

```
Run set -e
  set -e
  kubectl rook-ceph --operator-namespace test-operator -n test-cluster health
  shell: /usr/bin/bash --noprofile --norc -eo pipefail -x {0}
  env:
    ROOK_PLUGIN_SKIP_PROMPTS: true
+ set -e
+ kubectl rook-ceph --operator-namespace test-operator -n test-cluster health
Error: . %!!(string=failed to get rook version)w(<nil>)
Error: Process completed with exit code 1.
```

File: pkg/filesystem/subvolume.go [hunk 9]
```diff
@@ -149,7 +172,11 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
 	for _, subvolume := range subvollist {
 		check := checkStaleSubvolume(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs, subvolume, svg, k8sSubvolume)
 		if check {
-			exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvolume, svg}, OperatorNamespace, CephClusterNamespace, true, false)
+			_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvolume, svg}, OperatorNamespace, CephClusterNamespace, true)
+			if err != nil {
+				logging.Error(err, "failed to remove subvolume %q", subvolume)
```
Comment (subhamkrai) [inline pkg/filesystem/subvolume.go:177]:
```suggestion
				logging.Error(err, "failed to remove stale subvolume %q", subvolume)
```

File: pkg/mons/restore_quorum.go [hunk 10]
```diff
@@ -193,27 +193,42 @@ func updateMonMap(ctx context.Context, clientsets *k8sutil.Clientsets, clusterNa
...
+	_, err := exec.RunCommandInLabeledPod(ctx, clientsets, labelSelector, "mon", "ceph-mon", extractMonMapArgs, clusterNamespace, false)
+	if err != nil {
+		logging.Fatal(err, "failed to extract monmap")
+	}
 
 	logging.Info("Printing monmap")
-	exec.RunCommandInLabeledPod(ctx, clientsets, labelSelector, "mon", "monmaptool", []string{"--print", monmapPath}, clusterNamespace, false, true)
+	_, err = exec.RunCommandInLabeledPod(ctx, clientsets, labelSelector, "mon", "monmaptool", []string{"--print", monmapPath}, clusterNamespace, false)
+	if err != nil {
+		logging.Fatal(err, "failed to eprint monmap")
```
Comment (subhamkrai) [inline pkg/mons/restore_quorum.go:204]:
```suggestion
		logging.Fatal(err, "failed to print monmap")
```

File: pkg/filesystem/subvolume.go [hunk 11]
```diff
@@ -102,19 +113,27 @@ func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, o
...
+func getFileSystem(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string) ([]fsStruct, error) {
+	fsList, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "ls", "--format", "json"}, operatorNamespace, clusterNamespace, true)
+	if err != nil {
+		logging.Error(err, "failed to get filesystems")
+		return []fsStruct{}, err
+	}
 	fsstruct := unMarshaljson(fsList)
 	if len(fsstruct) == 0 {
-		logging.Fatal(fmt.Errorf("failed to get filesystem"))
+		logging.Fatal(fmt.Errorf("failed to get filesystems"))
```
Comment (subhamkrai) [inline pkg/filesystem/subvolume.go:124]:
```suggestion
		logging.Fatal(fmt.Errorf("no filesystem found"))
```

File: pkg/filesystem/subvolume.go [hunk 12]
```diff
@@ -124,13 +143,17 @@ func checkSnapshot(ctx context.Context, clientsets *k8sutil.Clientsets, operator
...
+func getSubvolumeGroup(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace, fs string) ([]fsStruct, error) {
+	svgList, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolumegroup", "ls", fs, "--format", "json"}, operatorNamespace, clusterNamespace, true)
+	if err != nil {
+		logging.Error(err, "failed to get subvolume groups for filesystem %q", fs)
+		return []fsStruct{}, err
+	}
 	subvolg := unMarshaljson(svgList)
 	if len(subvolg) == 0 {
-		logging.Fatal(fmt.Errorf("failed to get subvolumegroup for filesystem %q", fs))
+		logging.Fatal(fmt.Errorf("failed to get subvolumegroups for filesystem %q", fs))
```
Comment (subhamkrai) [inline pkg/filesystem/subvolume.go:154]:
```suggestion
		logging.Fatal(fmt.Errorf("no subvolumegroups found for filesystem %q", fs))
```

**Review by subhamkrai — CHANGES_REQUESTED** [review]
some nits. Thanks for the changes

File: pkg/mons/restore_quorum.go (hunk 10 above)
Comment (satoru-takeuchi) [reply pkg/mons/restore_quorum.go]: Oops, sorry for typo.

**Review by subhamkrai — APPROVED** [review]

**Pull request merged**